### ISSUE TITLE
Compaction, simplify, IF-collapse and restore change a creature's structure without deleting its uuid — a stale identity lets Fitness copy a score the creature never earned

### ISSUE THREAD

**Issue by nleck** [stSoftwareAU/NEAT-TS#3843]
## The rule

A creature's `uuid` is content-derived. Any structural change — neuron, synapse, bias or weight — must delete it so it is recomputed. `CreatureUtils.makeUUID` **short-circuits on an existing value and never recomputes**:

```ts
static makeUUID(creature: Creature): string {
  if (creature.uuid) {
    return creature.uuid;      // stale value survives forever
  }
```

So the invariant depends entirely on every mutation site remembering to `delete creature.uuid`. There are 26 sites that do.

## Five that do not

| file | uuid cleared? | memetic cleared? |
|---|---|---|
| `src/compact/CompactCreature.ts` | **no** | yes (`:103`, `:559`) |
| `src/compact/CompactUnused.ts` | **no** | — |
| `src/compact/IfCollapse.ts` | **no** | — |
| `src/optimize/Simplify.ts` | **no** | — |
| `src/blackbox/RestoreSource.ts` | **no** | — |
| `src/architecture/…/DiscoveryNeuronRemoval.ts` | yes (2 sites) | — |

Every one of the first five rewrites structure. `CompactCreature` deletes `memetic` on the way through, so the need to shed stale metadata was understood — `uuid` was just missed.

## Why a stale uuid is not cosmetic

`Fitness.ts:186-190` deduplicates the evaluation queue **by uuid**, and then copies the result across:

```ts
// Issue #1016: Deduplicate by UUID to avoid redundant evaluations
const duplicates = new Map<string, Creature[]>();
…
for (const duplicate of dupes) {
  if (duplicate !== creature) {
    duplicate.score = creature.score;      // never evaluated itself
    if (errorTag) addTag(duplicate, "error", errorTag);
    if (scoreTag) addTag(duplicate, "score", scoreTag);
  }
}
```

Combined with `Fitness.ts:184` — `population.filter((c) => c.score === undefined)` — a creature that arrives already carrying a score is never evaluated at all.

So a compacted or simplified creature that keeps its pre-change uuid can be treated as a duplicate of the creature it was derived from and **handed that creature's score without ever being scored**. That is precisely the failure mode of "trusting a score NEAT did not compute itself", reached by the back door.

It also compounds NEAT-AI#3840: those passes were already capable of changing behaviour while advertising exactness. A damaged creature that additionally inherits the healthy creature's score is invisible — it looks like a fold that cost nothing.

## Ask

1. Delete `uuid` in all five paths at the point structure changes.
2. Better than five fixes: make it structural. Either have `makeUUID` recompute when a cheap structure fingerprint disagrees with the cached value, or centralise mutation behind a helper that clears identity — so the next pass added cannot forget.
3. Regression test: mutate a creature (add/remove a neuron, change a weight) and assert `makeUUID` returns a different value afterwards; and assert two structurally different creatures never share a uuid.

Reported by the repo owner: *"The uuid must be truly unique. It's calculated — if there's any creature that isn't identical with the same UUID then that's a big bug."*

**Comment by nleck** [comment]
Auditing the Rust siblings against this rule prompted a full sweep of the TypeScript for the same shape. **Five additional sites** beyond the five already listed here, plus one primitive-level gap that is the natural structural fix point. Every line below was re-read and verified; confidence is stated per item.

First, a scoping note that removes a large class of false positives: `uuid` lives on `CreatureInternal` (`architecture/CreatureInterfaces.ts:48`), **not** on `CreatureExport`, which extends `CreatureCommon` and has no uuid field (`:65-70`). So the export-based compaction passes — `ConstantFold`, `ConstantMerge`, `DeadSubgraphPruning`, `AggressivePrune`, `SynapsePruning`, `SimplifyLargeWeights`, `OrphanedNeuronCleanup` — cannot carry a stale uuid through their own signature, and I am **not** reporting them. The sites below all operate on a live `Creature`.

---

### 1. `NEAT/NeatEvolution.ts:520-542` — the creative-thinking elitist clone

```ts
const creativeThinking = n.shallowClone();
delete creativeThinking.memetic;
delete creativeThinking.score;
const addConnection = new AddConnection(creativeThinking);
for (…) { addConnection.mutate(…); }
newPopulation.push(creativeThinking);
```

`shallowClone` copies the parent's identity verbatim — `creature/CreatureSerialization.ts:1049`, `clone.uuid = creature.uuid;`. `AddConnection.mutate` appends synapses, which are hash inputs. `mutate/AddConnection.ts` contains **zero** occurrences of `uuid`; it clears only `memetic` (`:149`, `:223`).

This is the one place the usual safety net does not apply. Every other mutation operator is reached through `Mutator.getMutatorInstance()` → `mutateCreature`, which clears the uuid at `NEAT/Mutator.ts:989`. This is the codebase's only direct instantiation of an operator, so the elitist's structurally-modified clone enters `newPopulation` wearing the elite parent's uuid. **Confidence: high.**

### 2. `architecture/Offspring.ts:628` and `:638` — repairs that run after the uuid is materialised

```
:576   delete offspring.uuid;
:577   const childUUID = CreatureUtil.makeUUID(offspring);   // identity fixed here
…
:628   Offspring.repairForwardOnlyComputationalInbounds(offspring, mother, father);
:638   Offspring.repairOrphanedConstants(offspring, mother, father);
```

Both helpers add synapses to the live offspring (`:866`, `:875`, `:937`, `:953` — `offspring.connect(...)`) and each ends with `offspring.clearCache()` only (`:879`, `:959`). `clearCache()` invalidates `topologyHash` and `_cachedUuidLookup` but deliberately never touches `uuid`. `Offspring.ts` contains exactly one `delete …uuid`, at `:576`, and there is no `fix()` after `:638`.

So the returned offspring's uuid describes the pre-repair topology — and that value has already been used for the "no point returning clones" check at `:609`. The ordering is correct for `RestoreGraftAliases` (`:570`, before the delete); only the post-`:577` repairs are the gap. **Confidence: high.**

### 3. `architecture/training/TrainingSetup.ts:200-206` — synthetic synapses added after the uuid is taken

```ts
if (options.syntheticSynapses) {
  const synResult = generateSyntheticSynapses(creature);
  syntheticKeys = synResult.syntheticKeys;
  // Invalidate WASM/state caches after structural change.
  creature.clearState();
}
```

`generateSyntheticSynapses` ends in `creature.connectBatch(connections)` (`propagate/SyntheticSynapses.ts:164`) — potentially thousands of new synapses. The comment names it a structural change, but only `clearState()` follows, and neither that nor `connectBatch` touches `uuid`. `TrainingSetup.ts` has **zero** occurrences of `uuid`.

The uuid is already materialised by then: `architecture/Training.ts:134` calls `CreatureUtil.makeUUID(creature)` and `:137` calls `prepareTraining(...)`. The stale value is live for the whole run — it is used as a filename at `TrainingOutcome.ts:131`. **Confidence: high.**

### 4. `upgrade/Upgrade.ts:84` — IF repair on the feedback-enabled branch

```ts
if (error.reason === "IF_CONDITIONS") {
  repairInvalidIfNeuronsInCreature(creature);
  pruneOrphanMemeticReferences(creature);
  try { creatureValidate(creature); return; }   // returns with the old uuid
```

`repairInvalidIfNeuronsInCreature` (`architecture/RepairInvalidIfNeurons.ts:38-50`) calls `neuron.setSquash("IDENTITY")` and deletes `syn.type` — **both are hash inputs** (`CreatureUtils.ts:115-131`). It clears `memetic` and calls `clearCache()`, so it plainly knows it is structural, but never clears `uuid`; `Neuron.setSquash` does not compensate.

Its other three callers are covered — `CreatureMutation.ts:273` (inside `fix()`), `mutate/SubNeuron.ts:101` (via `Mutator`), `Upgrade.ts:121` (followed immediately by `fix()`). This branch is not: `upgrade/Upgrade.ts` contains **zero** `delete …uuid`, and it runs on live uuid-bearing creatures via `prepareCreatureForBreeding()` on a `shallowClone`d parent. **Confidence: high.**

### 5. `creature/CreatureTraining.ts:250` — the invalidation is gated on the wrong condition

```ts
propagateUpdate(creature, config, sparseConfig);   // writes every weight and bias
let changed = false;
… changed ||= n.applyLearnings(); …
if (changed) { delete creature.uuid; delete creature.memetic; … }
```

`propagateUpdate` writes weights and biases throughout — `neuron/NeuronPropagation.ts:63, 69, 121, 127, 134, 140` (`c.weight = clampAndTrack(...)`, `neuron.bias = clampAndTrack(...)`). Both are hash inputs. But `changed` is set **only** by `Neuron.applyLearnings()`, which reports an IF/MAXIMUM/MINIMUM structural downgrade — not whether the gradient step moved anything. A normal backprop pass moves every weight and leaves `changed === false`, so the uuid survives a full training step.

Worth contrasting with `predictiveCoding/PredictiveCodingTrainer.ts:272-273`, which does the same class of Hebbian/gradient write and unconditionally clears both fields. That is the correct shape; this is the same operation missing it. **Confidence: high on the code fact.**

---

### The structural fix point

Two shared primitives clear `memetic` but never `uuid`:

- `removeHiddenNeuron()` — `compact/OrphanedNeuronCleanup.ts:136-139`, `delete creature.memetic;` and nothing else. **13 call sites.**
- the whole `compact/MemeticCleanup.ts` family (`cleanupMemeticForRemovedSynapse`, `cleanupMemeticForRemovedNeuron`, `pruneOrphanMemeticReferences`) — memetic only, by name and by behaviour.

Every current `removeHiddenNeuron` caller happens to be covered one level up, so this is latent rather than live — but it is the reason the five sites in this issue are scattered rather than singular. The two genuine safe harbours are `Creature.fix()` (`creature/CreatureMutation.ts:285`) and `loadFrom()`; `clearCache()`, `clearState()`, `connect`/`disconnect`/`connectBatch` and `Neuron.setSquash()` all look like invalidation points and are not.

Rather than adding a sixth, seventh and eighth scattered `delete`, consider making the two removal/mutation primitives clear `uuid` alongside `memetic`, and treating any structural edit that reaches neither `fix()` nor `loadFrom()` as the thing to assert against. `CreatureMutation.ts:283-289` already demonstrates the robust idiom — delete, recompute, and compare — rather than trusting a boolean:

```ts
delete creature.uuid;
const endUUID = CreatureUtil.makeUUID(creature);
if (startUUID !== endUUID) { delete creature.memetic; }
```

Site 5 above is precisely a case where a hand-maintained `changed` flag drifted out of sync with what actually changed; the recompute-and-compare pattern cannot drift.

### TDD

Beyond the per-site red tests, one table-driven test would cover the class and catch the next instance: for each structural entry point, snapshot `makeUUID` before, apply, then assert the recomputed uuid **differs**. Pair it with the inverse guard — a pass that changes only `tags` must leave the uuid **unchanged**, since tags are excluded from the hash — and with an assertion that per-neuron `uuid` values are preserved, because they are provenance keys and an *input* to the creature hash rather than something to clear.

### Rule 2 (#3844) — no additional sites

The same sweep found **no** additional module that removes a neuron or synapse without clearing or pruning `memetic`. Two that looked like hits are covered: `compact/ParallelBridgeMerge.ts:291-299` and `compact/RemoveBackwardSynapses.ts:60-62` have no memetic handling of their own, but their sole caller sets the flag that reaches `delete compactCreature.memetic` at `CompactCreature.ts:559`. The commented-out `// delete this.creature.memetic;` at `mutate/AddNeuron.ts:288` is **correct** — that operator only splices in a neuron and connects; it never disconnects, so it is append-only and existing memetic keys stay valid.

**Comment by stservice** [comment]
✅ **Vibe Coder released this claim** — host `vibe-coder-13083`, finished 02:10 UTC. Superseded by closed #3846 — https://github.com/stSoftwareAU/NEAT-AI/pull/3846 (this run raised no PR). WIP preserved: 3 checkpoint commits pushed to 'issue-3843-stale-uuid-after-structural-change' — the next claim resumes from that branch (Issue #4170).
